File: AlgoExpert/tournament-winner.go
package main

const HOME_TEAM_WON = 1
const WIN_POINTS = 3

func TournamentWinner(competitions [][]string, results []int) string {
	pointMap := make(map[string]int)
	totalMatches := len(results)
	result := ""

	for i := 0; i < totalMatches; i++ {
		homeTeam, awayTeam := competitions[i][0], competitions[i][1]
		winner := awayTeam
		if HOME_TEAM_WON == results[i] {
			winner = homeTeam
		}
		winnerPoints := pointMap[winner] + WIN_POINTS
		pointMap[winner] = winnerPoints

		if pointMap[result] < winnerPoints {
			result = winner
		}
	}
	return result
}

File: AlgoExpert/tournament-winner.ts
export function tournamentWinner(competitions: string[][], results: number[]) {
  const gameResults: { [key: string]: number } = {};
  let currentMax = 0,
    currentTopTeam = "";
  for (const [index, [home, away]] of competitions.entries()) {
    gameResults[home] = gameResults[home] || 0;
    gameResults[away] = gameResults[away] || 0;
    let winner = away;
    if (results[index] === 1) {
      winner = home;
    }
    gameResults[winner] += 3;
    const winnerPoints = gameResults[winner];
    if (winnerPoints > currentMax) {
      currentMax = winnerPoints;
      currentTopTeam = winner;
    }
  }

  return currentTopTeam;
}
